iocuddle/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+    ///
+    /// No argument is supplied to the internal `ioctl()` call. The raw
+    /// (positive) return value from the internal `ioctl()` call is returned
+    /// on success.
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl<T> Ioctl<Read, &T> {
+    /// Issue an `ioctl` to read a file descriptor's metadata as type `T`.
+    ///
+    /// A zeroed instance of type `T` is passed as the first argument to the
+    /// internal `ioctl()` call. Upon success, returns the raw (positive)
+    /// return value and the instance of `T`.
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<(c_uint, T)> {
+        let mut out: T = unsafe { core::mem::MaybeUninit::zeroed().assume_init() };
+
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, &mut out as *mut _, null::<c_void>()) };
+
+        r.try_into()
+            .and_then(|x| Ok((x, out)))
+            .or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl Ioctl<Write, c_void> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// No argument is provided.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl Ioctl<Write, c_int> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// A C-integer argument is provided.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd, data: c_int) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, data, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl<T> Ioctl<Write, &T> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// A reference to an immutable instance of `T` is provided as the argument.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd, data: &T) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, data as *const _, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl<T> Ioctl<WriteRead, &T> {
+    /// Issue an `ioctl` to modify a file descriptor and read its metadata
+    ///
+    /// A reference to a mutable instance of `T` is provided as the argument.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd, data: &mut T) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, data as *mut _, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
There isn't a test for the `Ioctl<_, c_void>` impls, are they considered unsupported at this point?

Reply:
I don't know how to test them. Unfortunately, ioctls are usually device specific. The only simple device we have access to in the test suite is `/dev/tty`, which has no ioctls without arguments. There are a very large number of ioctls out there. So I'm sure there is one we could use for the test. But there isn't an easy way to find that information.

Reply:
I see, I had no luck finding one quickly.